### PR TITLE
Add Host Observer

### PR DESCRIPTION
**Pull request by asuresh4**
Depends on #427 

**Description:** Add host observer to look for listening network endpoints on host

**Testing:** Unit tests added. Manually tested with `receivercreator` receiver.

**Documentation:** [README.md](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/eabd8d4fe646d7589466c471dc27f96bc0d7c530/extension/observer/hostobserver/README.md).

### CODE REVIEW

File: extension/observer/hostobserver/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+# Host Observer Extension
+
+**Status: beta**
+
+Looks at the current host for listening network endpoints. It will look for all listening sockets on TCP and UDP over IPv4 and IPv6.
```
Comment (keitwb) [inline extension/observer/hostobserver/README.md:5]: Add a subject to the first sentence.

File: extension/observer/hostobserver/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+It uses the /proc filesystem and requires the SYS_PTRACE and DAC_READ_SEARCH capabilities so that it can determine what processes own the listening sockets.
+
+### Configuration
+
+#### `refresh_interval`
+
+Determines how to look for changes in endpoints.
```
Comment (keitwb) [inline extension/observer/hostobserver/README.md:13]: how often?

File: extension/observer/hostobserver/README.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+Endpoint variables exposed by this observer are as follows.
+
+| Variable  | Description                                                                                |
+|-----------|--------------------------------------------------------------------------------------------|
+| type.port | `true`                                                                                     |
+| name      | name of the process associated to the port                                                 |
+| port      | port number                                                                                |
+| command   | full command used to invoke this process, including the executable itself at the beginning |
+| is_ipv6   | `true` if the endpoint is IPv6                                                             |
+| protocol  | "TCP" or "UDP"                                                                             |
```
Comment (keitwb) [inline extension/observer/hostobserver/README.md:28]: This is really the transport protocol and IPv4 vs v6 is the network protocol.  That distinction was not made very clear in the Smart Agent host observer.

File: extension/observer/hostobserver/extension.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+		isTCPOrUDP := c.Type == syscall.SOCK_STREAM || c.Type == syscall.SOCK_DGRAM
+		// UDP doesn't have any status
+		isUDPOrListening := c.Type == syscall.SOCK_DGRAM || c.Status == "LISTEN"
+		// UDP is "listening" when it has a remote port of 0
+		isTCPOrHasNoRemotePort := c.Type == syscall.SOCK_STREAM || c.Raddr.Port == 0
+
+		// PID of 0 means that the listening file descriptor couldn't be mapped
+		// back to a process's set of open file descriptors in /proc
+		if !isIPSocket || !isTCPOrUDP || !isUDPOrListening || !isTCPOrHasNoRemotePort || c.Pid == 0 {
+			continue
```
Comment (keitwb) [inline extension/observer/hostobserver/extension.go:90]: It might be useful behavior to still discover endpoints without the extra Caps and just omit the process information from the output.  Some people might just want to know if a port is listening and not care about process metadata.

File: extension/observer/hostobserver/extension.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+var _ component.ServiceExtension = (*hostObserver)(nil)
+
+func newObserver(logger *zap.Logger, config *Config) (component.ServiceExtension, error) {
+	h := &hostObserver{
+		logger: logger,
+		EndpointsWatcher: observer.EndpointsWatcher{
+			RefreshInterval: config.RefreshInterval,
+			ListEndpoints:   (&hostObserver{}).listEndpoints,
```
Comment (keitwb) [inline extension/observer/hostobserver/extension.go:43]: This is pretty ugly creating a completely separate instance of the host observer to pass the listEndpoints method to.  Why don't you just make listEndpoints a plain function instead?

Comment (keitwb) [reply extension/observer/hostobserver/extension.go:43]: Also logger would be `nil` in listEndpoints which is not going to behave properly.

**Review by keitwb — CHANGES_REQUESTED** [review]

File: extension/observer/hostobserver/README.md (hunk 3 above)
Comment (asuresh4) [reply extension/observer/hostobserver/README.md:28]: Would you rather have endpoint variables called `transport_protocol` and `network_protocol` instead of `protocol` and `is_ipv6`?

File: extension/observer/hostobserver/extension.go (hunk 4 above)
Comment (asuresh4) [reply extension/observer/hostobserver/extension.go:90]: I'm not sure I follow. Could you please clarify?

File: extension/observer/hostobserver/extension.go (hunk 5 above)
Comment (asuresh4) [reply extension/observer/hostobserver/extension.go:43]: Will refactor this and also make sure there's test coverage around this!

File: extension/observer/hostobserver/README.md (hunk 3 above)
Comment (asuresh4) [reply extension/observer/hostobserver/README.md:28]: Discussed with @jrcamp as well and we were thinking of updating these variables to `transport_protocol` (either `TCP` or `UDP`) and `network_protocol` (either `ipv4` or `ipv6`). What do you think @keitwb ?

File: extension/observer/hostobserver/extension.go (hunk 4 above)
Comment (keitwb) [reply extension/observer/hostobserver/extension.go:90]: What happens if you don't have the right caps on Linux is that c.Pid will be 0, and thus you will completely ignore this listening socket.  But some people might want to know that the socket is listening and do discovery rules on them even though they don't get any of the process information.

File: extension/observer/hostobserver/README.md (hunk 3 above)
Comment (asuresh4) [reply extension/observer/hostobserver/README.md:28]: ptal @jrcamp @keitwb - https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/473

File: go.mod [hunk 6]
```diff
@@ -7,34 +7,35 @@ require (
 	github.com/golangci/golangci-lint v1.28.3
 	github.com/google/addlicense v0.0.0-20200622132530-df58acafd6d5
 	github.com/jstemmer/go-junit-report v0.9.1
-	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/alibabacloudlogserviceexporter v0.0.0-00010101000000-000000000000
```
Comment (rmfitzpatrick) [inline go.mod:10]: Were these deviations from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/487/files#diff-37aff102a57d3d7b797f152915a6dc16 intentional?

Comment (jrcamp) [reply go.mod:10]: @bogdandrutu do you know why these were all changed ^

File: extension/observer/hostobserver/extension.go [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+			e.logger.Error(
+				"Skipping process",
+				zap.Int32("pid", pid),
+				zap.Error(err),
+			)
+			continue
+		}
+
+		for _, c := range conns {
+			cd := collectConnectionDetails(*c)
```
Comment (jrcamp) [inline extension/observer/hostobserver/extension.go:146]: If it's already a pointer probably better to just pass as a pointer instead of making a copy.

File: extension/observer/endpoints.go [hunk 8]
```diff
@@ -50,6 +50,15 @@ type Port struct {
 	Transport Transport
 }
 
+// HostPort is an endpoint discovered on a host.
+type HostPort struct {
+	Name      string
```
[the remark this answers is not in the document]
Comment (asuresh4) [reply extension/observer/endpoints.go]: Added descriptions to both, fields in `HostPort` and `Port`.

File: go.mod (hunk 6 above)
Comment (asuresh4) [reply go.mod:10]: Yes, these were intentional. Typically we just have the version be `v0.0.0` for modules in this repo. Not sure why it was updated. But @bogdandrutu 's changes are as desired, I'll leave it unchanged in this diff.

File: extension/observer/endpoints.go [hunk 9]
```diff
@@ -44,10 +44,32 @@ type Pod struct {
...
+}
+
+// HostPort is an endpoint discovered on a host.
+type HostPort struct {
+	// Name of the process associated to Endpoint.  If host_observer
+	// is unable to collect information about process using the
+	// Port, this value is set by the observer to be "IP-Port-Transport"
+	// prefixed by the name of the host_observer instance,
+	// "(host_observer/2)127.0.0.1-9099-UDP" for example.
+	Name string
```
Comment (jrcamp) [inline extension/observer/endpoints.go:64]: If we can't collect it should we just make name be empty string? Why set it to id?

File: extension/observer/hostobserver/extension.go [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+
+		// PID of 0 means that the listening file descriptor couldn't be mapped
+		// back to a process's set of open file descriptors in /proc. Collect these
+		// endpoints even though there's no process metadata available so users can
+		// still do discovery rules on such sockets.
+		if c.Pid == 0 {
+			cd := collectConnectionDetails(&c)
+			id := observer.EndpointID(
+				fmt.Sprintf(
+					"(%s)%s-%d-%s", e.observerName, c.Laddr.IP, cd.port, cd.transport,
```
Comment (jrcamp) [inline extension/observer/hostobserver/extension.go:112]: I'd like to make it and k8sobserver id format similar but there's some tweaks I want to make to id so I'll just do that separately in another PR.

I think it's better if we use metadata on the end of the ID, similar to what receiver_creator does:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/bb81db76bc6130e348ab6b2ab401d328b537f4ac/receiver/receivercreator/runner.go#L101

So this would look something like:

`hostobserver/2{endpoint="127.0.0.1:23", protocol="UDP", pid=3482}`

or something like that.

(Just as an aside RC should probably just use the endpoint ID to resolve that TODO)

tl;dr: nothing for you to change here right now, just notes on things to fix sometime.

File: extension/observer/endpoints.go (hunk 9 above)
Comment (asuresh4) [reply extension/observer/endpoints.go:64]: That sounds good. Will update.

**Review by jrcamp — APPROVED** [review]

**Review by jrcamp — CHANGES_REQUESTED** [review]
Just noticed, look like host observer test failed due to races.

**Comment by asuresh4** [comment]
> Just noticed, look like host observer test failed due to races.

Looks like the race was from a couple of tests where I tried to mock failures of third-party packages. Removing those tests for now. Let me know if you feel strongly about adding those back.

**Comment by asuresh4** [comment]
```
make[2]: Entering directory '/home/circleci/project/receiver/signalfxreceiver'
running go unit test ./... + coverage in /home/circleci/project/receiver/signalfxreceiver
--- FAIL: Test_signalfxeceiver_EndToEnd (0.01s)
    receiver_test.go:142: 
        	Error Trace:	receiver_test.go:142
        	Error:      	Received unexpected error:
        	            	Post "http://127.0.0.1:45585/v2/datapoint": dial tcp 127.0.0.1:45585: connect: connection refused
        	Test:       	Test_signalfxeceiver_EndToEnd
FAIL
coverage: 95.8% of statements
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/signalfxreceiver	0.578s
```

Test failure seems unrelated to this PR, will try a rebase.

**Comment by asuresh4** [comment]
ptal @jrcamp

**Review by jrcamp — APPROVED** [review]

File: extension/observer/hostobserver/extension.go [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+		return nil, fmt.Errorf("could not get process args: %v", err)
+	}
+
+	return &processDetails{
+		name: name,
+		args: args,
+	}, nil
+}
+
+var getProcessName = func(proc *process.Process) (string, error) {
```
Comment (jrcamp) [inline extension/observer/hostobserver/extension.go:236]: Maybe in followup PR you could try refactoring this into a struct and/or interface that can be injected by a mock implementation for tests. Should be safer than trying to modify package level variables in test. Can talk about it more offline if it's not clear what I mean.

Comment (asuresh4) [reply extension/observer/hostobserver/extension.go:236]: I think understand what you mean. Will make a subsequent PR to refactor.

**Comment by bogdandrutu** [comment]
Please rebase

**Comment by asuresh4** [comment]
@bogdandrutu done - ready to merge when you are.
